src/air/constraints/evaluator.rs

```diff
@@ -153,16 +153,20 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
             })
             .collect();
 
-        let x_n = Polynomial::new_monomial(FieldElement::<F>::one(), trace_length);
-        let x_n_1 = x_n - FieldElement::<F>::one();
+        let blowup_factor_log2 = u64::from(blowup_factor.trailing_zeros());
 
-        let mut zerofier_evaluations = evaluate_polynomial_on_lde_domain(
-            &x_n_1,
-            domain.blowup_factor,
-            domain.interpolation_domain_size,
-            &domain.coset_offset,
+        let offset = FieldElement::<F>::from(self.air.context().options.coset_offset);
+        let offset_pow = offset.pow(trace_length);
+        let one = FieldElement::<F>::one();
+        let mut zerofier_evaluations = get_powers_of_primitive_root_coset(
```

Comment:
For a different PR, but I'd suggest `get_powers_of_primitive_root_coset` to return the iterator. Or, at least, to implement an `iter_powers_of_primitive_root_coset` and implement the former in terms of that and a `collect`. This avoids creating a `Vec<_>` that we only iterate once.

Reply:
We can do that in another PR